src/validation/permission-exclusions.ts

```diff
@@ -0,0 +1,303 @@
+ ******/
+
+import * as keto from '@ory/keto-client'
+import { ValidationError } from './validation-error'
+import { ServiceConfig } from '../shared/config'
+import { logger } from '../shared/logger'
+
+export interface UserRole {
+  username: string;
+  roles: string[];
+}
+
+export interface RolePermissions {
+  rolename: string;
+  permissions: string[];
+}
+
+export interface PermissionExclusions {
+  permissionsA: string[];
+  permissionsB: string[];
+}
+
+export interface PermissionExclusionCombos {
+  permissionA: string;
+  permissionB: string;
+}
+
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+export function isPermissionExclusionCombos (object: PermissionExclusions | PermissionExclusionCombos): boolean {
+  return 'permissionA' in object
+}
+
+export class PermissionExclusionsValidator {
+  oryKetoReadApi: keto.ReadApi
+  serviceConfig: ServiceConfig
+
+  constructor (serviceConfig: ServiceConfig) {
+    this.serviceConfig = serviceConfig
+    this.oryKetoReadApi = new keto.ReadApi(
+      undefined,
+      serviceConfig.ORY_KETO_READ_SERVICE_URL
+    )
+  }
+
+  // // Commenting the following because this logic is not being used anywhere and user permission validation is done
+  // // a different way
+  // // -------------------
+  // async validateUserPermissions (userId: string, permissionsToAssign: Set<string>) : Promise<void> {
+  //   const validationErrors : string[] = []
+  //   // Iterate though the list of new permissions and find out if it conflicts with other permissions of the user
+  //   // logger.info(JSON.stringify(newPermissions, null, 2))
+  //   const newPermissions = Array.from(permissionsToAssign.values())
+  //   let validationFailed = false
+  //   for (let i = 0; i < newPermissions.length; i++) {
+  //     try {
+  //       const checkPermissionsExclusionsResponse = await this.oryKetoReadApi.getCheck(
+  //         'permission',
+  //         newPermissions[i],
+  //         'excludes',
+  //         userId
+  //       )
+  //       const permissionExclusionClash = checkPermissionsExclusionsResponse.data?.allowed || false
+  //       if (permissionExclusionClash) {
+  //         validationErrors.push(`ERROR: permission '${newPermissions[i]}' can not be assigned`)
+  //         validationFailed = true
+  //       }
+  //     } catch (err) {}
+  //   }
+  //   logger.info(validationErrors)
+  //   if (validationFailed) {
+  //     throw new ValidationError(validationErrors)
+  //   }
+  //   // TODO: We can stop the validation when the first error occurs based on a config param like quickCheck
+  // }
+
+  _getPermissionExclusionsForPermission (
+    permission: string,
+    permissionExclusions: PermissionExclusions[] | PermissionExclusionCombos[]
+  ) : Set<string> {
+    const permissionExclusionsSet: Set<string> = new Set()
+    if (permissionExclusions.length <= 0) {
+      return permissionExclusionsSet
+    }
+    if (isPermissionExclusionCombos(permissionExclusions[0])) {
```

Comment:
_If_ my earlier change to `isPermissionExclusionCombos` is accepted:
```suggestion
    if (isPermissionExclusionCombos(permissionExclusions)) {
```